mlflow/gateway/config.py

```diff
@@ -0,0 +1,277 @@
+import pathlib
+from enum import Enum
+import json
+import logging
+import os
+from pathlib import Path
+from pydantic import BaseModel, validator, parse_obj_as, Extra, ValidationError
+from pydantic.json import pydantic_encoder
+from typing import Optional, Union, List, Dict, Any
+import yaml
+
+from mlflow.exceptions import MlflowException
+from mlflow.gateway.constants import PROVIDERS
+from mlflow.gateway.utils import is_valid_endpoint_name, check_configuration_route_name_collisions
+
+
+_logger = logging.getLogger(__name__)
+
+
+class Provider(str, Enum):
+    CUSTOM = "custom"
+    OPENAI = "openai"
+    ANTHROPIC = "anthropic"
+    DATABRICKS_SERVING_ENDPOINT = "databricks_serving_endpoint"
+    MLFLOW = "mlflow"
+
+
+class RouteType(str, Enum):
+    CUSTOM = "custom"
+    LLM_V1_COMPLETIONS = "llm/v1/completions"
+    LLM_V1_CHAT = "llm/v1/chat"
+
+
+class OpenAIConfig(BaseModel, extra=Extra.forbid):
+    openai_api_key: Optional[str] = None
+    openai_api_type: Optional[str] = None
+    openai_api_base: Optional[str] = "https://api.openai.com/v1"
+    openai_api_version: Optional[str] = None
+    openai_organization: Optional[str] = None
+
+
+class AnthropicConfig(BaseModel, extra=Extra.forbid):
+    anthropic_api_key: Optional[str] = None
+    anthropic_api_base: Optional[str] = "https://api.anthropic.com/"
+
+
+class DatabricksConfig(BaseModel, extra=Extra.forbid):
+    databricks_api_token: Optional[str] = None
+    databricks_api_base: str
+
+
+class MLflowConfig(BaseModel, extra=Extra.forbid):
+    api_base: str
+
+
+class CustomConfig(BaseModel, extra=Extra.forbid):
+    api_key: Optional[str] = None
+    api_base: str
+    api_version: Optional[str] = None
+
+
+config_types = {
+    Provider.OPENAI: OpenAIConfig,
+    Provider.ANTHROPIC: AnthropicConfig,
+    Provider.DATABRICKS_SERVING_ENDPOINT: DatabricksConfig,
+    Provider.MLFLOW: MLflowConfig,
+    Provider.CUSTOM: CustomConfig,
+}
+
+
+class ModelInfo(BaseModel, extra=Extra.forbid):
+    name: Optional[str] = None
+    provider: Provider = Provider.CUSTOM
+
+
+def _resolve_api_key_from_input(api_key_input):
+    """
+    Resolves the provided API key.
+
+    Input formats accepted:
+
+    - Path to a file as a string which will have the key loaded from it
+    - environment variable name that stores the api key
+    - the api key itself
+    """
+
+    if not isinstance(api_key_input, str):
+        raise MlflowException.invalid_parameter_value(
+            "The api key provided is not a string. Please provide either an environment "
+            "variable key, a path to a file containing the api key, or the api key itself"
+        )
+
+    # try reading as an environment variable
+    env_var_attempt = api_key_input[1:] if api_key_input.startswith("$") else api_key_input
+
+    env_var = os.getenv(env_var_attempt)
+    if env_var:
+        return env_var
```

Comment:
I want to make sure the logig here is correct. In the current logic, `openai_api_key: OPENAI_API_KEY` means `use the env var OPENAI_API_KEY`.

Reply:
Yes, I agree; we should only use an env var when the key starts with a dollar sign

Reply:
Filed #8710